processing-service-core/api/models/tenant.js

```diff
@@ -0,0 +1,340 @@
+}
+
+/**
+ * Creates or updates tenant specific configs.
+ * Creates or updates a custom configuration setting for this key against the tenant.
+ *
+ * @param {String} tenantId  String The ID of the tenant.
+ * @param {String} tenantConfigs TenantConfigs The json string representation key-value pair map of custom tenant configuration.
+ * no response value expected for this operation
+ **/
+function setTenantConfigs(tenantId, tenantConfigs) {
+    var tenantConfig = Q.defer();
+    var promiseArray = [];
+
+    for (var i = 0; i < tenantConfigs.configurations.length; i++) {
+        var config = tenantConfigs.configurations[i];
+        promiseArray.push(tenantConfigStoreModel.setTenantConfig(tenantId, config.key, config.value));
+    }
+    Q.all(promiseArray)
+            .then(function (results) {
+                logger.debug("Created tenant specific config for tenant " + tenantId);
+                tenantConfig.resolve();
+            })
+            .fail(function (errorResponse) {
+                logger.debug("Unable to create tenant specific config for tenant " + tenantId);
+                tenantConfig.reject(errorResponse);
+            })
+            .done();
+    return tenantConfig.promise;
+}
+
+/**
+ * Delete tenant specific config for the provided key.
+ * Deletes tenant specific configuration for the key that has been provided.
+ *
+ * @param {String} tenantId String The ID of the tenant.
+ * @param {String} key String The custom configuration key.
+ * no response value expected for this operation
+ **/
+function deleteTenantConfig(tenantId, key) {
+    var tenantConfig = Q.defer();
+    logger.debug("Deleting tenant specific config for tenant " + tenantId + " with key " + key);
+    tenantConfigStoreModel.deleteTenantConfig(tenantId, key)
+            .then(function (deleteConfigPromise) {
+                logger.debug("Successfully deleted tenant specific config for tenant " + tenantId + " with key " + key);
+                tenantConfig.resolve(deleteConfigPromise);
+            })
+            .fail(function (errorResponse) {
+                logger.debug("Failed to delete tenant specific config for tenant " + tenantId + " with key " + key);
+                tenantConfig.reject(errorResponse);
+            })
+            .done();
+    return tenantConfig.promise;
+}
+
+/**
+ * Delete tenant specific configs for the provided tenantId.
+ * Deletes all tenant specific configuration for tenant that has been provided.
+ *
+ * @param {String} tenantId String The ID of the tenant.
+ * no response value expected for this operation
+ **/
+function deleteTenantConfigs(tenantId) {
+    var tenantConfig = Q.defer();
+    tenantConfigStoreModel.getTenantConfigs(tenantId)
+            .then(function (tenantConfigs) {
+                var tenantConfigsMap = createMap(tenantConfigs);
+                for (var key in tenantConfigsMap) {
+                    tenantConfigStoreModel.deleteTenantConfig(tenantId, key);
```

Comment:
This returns a promise. You need to gather all the promises and return them from this method so the next step doesn't execute until all deletions have occurred.
An example of working with multiple promises created in a loop can be seen [here](https://github.com/CAFDataProcessing/data-processing-service/blob/develop/processing-service-core/api/models/action.js#L495).